### PR TITLE
[SPARK-18872][SQL][TESTS] New test cases for EXISTS subquery (Aggregate, Having, Orderby, Limit)

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
This PR adds the second set of tests for EXISTS subquery.  

File name                        | Brief description
------------------------| -----------------
exists-aggregate.sql              |Tests aggregate expressions in outer query and EXISTS subquery.
exists-having.sql|Tests HAVING clause in subquery.
exists-orderby-limit.sql|Tests EXISTS subquery support with ORDER BY and LIMIT clauses.

DB2 results are attached here as reference :

[exists-aggregate-db2.txt](https://github.com/apache/spark/files/743287/exists-aggregate-db2.txt)
[exists-having-db2.txt](https://github.com/apache/spark/files/743286/exists-having-db2.txt)
[exists-orderby-limit-db2.txt](https://github.com/apache/spark/files/743288/exists-orderby-limit-db2.txt)

##  How the patch was tested.
The test result is compared with the result run from another SQL engine (in this case is IBM DB2). If the result are equivalent, we assume the result is correct.

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-aggregate.sql.out [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                   WHERE  emp.emp_name = bonus.emp_name 
+                          AND EXISTS (SELECT Max(dept.dept_id) 
+                                      FROM   dept 
+                                      WHERE  emp.dept_id = dept.dept_id 
+                                      GROUP  BY dept.dept_id))
+-- !query 10 schema
+struct<emp_name:string,bonus_amt:double>
+-- !query 10 output
+emp 5	1000.0
+emp 6 - no dept	500.0
```
Comment (nsyca) [inline sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-aggregate.sql.out]: I have compared the result set matched with the result from DB2.

File: sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-having.sql.out [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+               WHERE  dept.dept_id = dept_id 
+               GROUP  BY dept_id 
+               HAVING EXISTS (SELECT 1 
+                              FROM   bonus 
+                              WHERE  ( bonus_amt > min(emp.salary) 
+                                       AND count(emp.dept_id) > 1 )))
+-- !query 7 schema
+struct<dept_id:int,dept_name:string,state:string>
+-- !query 7 output
+10	dept 1	CA
```
Comment (nsyca) [inline sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-having.sql.out]: I have compared the result set matched with the result from DB2.

File: sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-orderby-limit.sql.out [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+-- !query 11 output
+100	emp 1	2005-01-01	100.0	10
+100	emp 1	2005-01-01	100.0	10
+200	emp 2	2003-01-01	200.0	10
+300	emp 3	2002-01-01	300.0	20
+400	emp 4	2005-01-01	400.0	30
+500	emp 5	2001-01-01	400.0	NULL
+600	emp 6 - no dept	2001-01-01	400.0	100
+700	emp 7	2010-01-01	400.0	100
+800	emp 8	2016-01-01	150.0	70
```
Comment (nsyca) [inline sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-orderby-limit.sql.out]: I have compared the result set matched with the result from DB2.

**Comment by dilipbiswal** [comment]
cc @hvanhovell @gatorsmile @nsyca 
@nsyca Thanks for reviewing.

**Comment by gatorsmile** [comment]
Did the comparison. The results match what DB2 returned. Do you think we can remove some duplicate test cases in `SubquerySuite.scala`?

**Review by gatorsmile — APPROVED** [review]
Did the comparison. The results match what DB2 returned. Do you think we can remove some duplicate test cases in SubquerySuite.scala?

**Comment by dilipbiswal** [comment]
@gatorsmile Thanks a lot for reviewing. @nsyca and i had a brief discussion about this. Yeah, we will remove the duplicate test cases from SubquerySuite.scala. However , we wanted to keep all the tests till the phase1 work is committed to ensure the full coverage i.e we don't inadvertently remove any tests for which we may have missed to add a corresponding test in sqltestquerysuite.

**Comment by gatorsmile** [comment]
@dilipbiswal Does that mean your test cases cover all the EXISTS scenarios in `SubquerySuite.scala`?

**Comment by dilipbiswal** [comment]
@gatorsmile Currently for exists we are missing 1) Negative scenarios 2) Generators (there is one test case in subquerysuite) for it.

**Comment by gatorsmile** [comment]
@dilipbiswal Are you planning to submit another PR for `Generators` or do it in this PR?

**Comment by dilipbiswal** [comment]
@gatorsmile Yeah Sean. Actually most likely i will need to work out a different schema than what i have currently for the generator tests. So i was planning to add the negative scenarios and generator tests in another PR.

**Comment by gatorsmile** [comment]
ok. LGTM

**Comment by gatorsmile** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

**Comment by nsyca** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by dilipbiswal** [comment]
Many thanks @gatorsmile

**Pull request closed without merging**